e2e/main.go

```diff
@@ -0,0 +1,339 @@
+		}
+
+		if err := d.StopContainers(); err != nil {
+			log.Fatalf("[txtdirect_e2e]: Couldn't stop containers: %s", err.Error())
+		}
+	}
+
+	if err := d.ExamineLogs(); err != nil {
+		log.Fatalf("[txtdirect_e2e]: Couldn't examine the logs and count the stats: %s", err.Error())
+	}
+
+	if d.stats.failed != 0 {
+		log.Fatalf("[txtdirect_e2e]: Tests failed because there were %d failed tests.", d.stats.failed)
+	}
+}
+
+func listDirectories(directories *[]string) error {
+	return filepath.Walk(".", func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return fmt.Errorf("Couldn't find the directory %q: %s", path, err.Error())
+		}
+		if info.IsDir() && path != "." {
+			*directories = append(*directories, path)
+			return filepath.SkipDir
+		}
+		return nil
+	})
+}
+
+// CreateClient creates a Docker client and context and attaches them to the
+// dockerManager instance
+func (d *dockerManager) CreateClient() error {
+	d.ctx = context.Background()
+	cli, err := client.NewClientWithOpts(client.WithVersion("1.39"))
+	if err != nil {
+		return fmt.Errorf("Couldn't start the Docker client: %s", err.Error())
+	}
+	d.cli = cli
+	return nil
+}
+
+// StartContainers starts a CoreDNS and a TXTDirect containers for each test-case
+func (d *dockerManager) StartContainers() error {
+	// Get current working directory to create mount test-case's data to containers
+	var err error
+	d.cdir, err = os.Getwd()
+	if err != nil {
+		return fmt.Errorf("Couldn't get the current working directory: %s", err.Error())
+	}
+
+	d.network, err = d.cli.NetworkCreate(d.ctx, "coretxtd", types.NetworkCreate{
+		IPAM: &network.IPAM{
+			Config: []network.IPAMConfig{
+				{
+					IPRange: "172.20.10.0/24",
+					Subnet:  "172.20.0.0/16",
+				},
+			},
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("Couldn't create the network adaptor: %s", err.Error())
+	}
+
+	// Create the CoreDNS container
+	d.cdContainer, err = d.cli.ContainerCreate(d.ctx, &container.Config{
+		Image: "coredns/coredns",
+		Cmd:   []string{"-conf", "/e2e/Corefile"},
+		ExposedPorts: nat.PortSet{
+			"53/tcp": struct{}{},
+			"53/udp": struct{}{},
+		},
+	}, &container.HostConfig{
+		Mounts: []mount.Mount{
+			{
+				Type:   mount.TypeBind,
+				Source: d.cdir + "/" + d.dir,
+				Target: "/e2e",
+			},
+		},
+		PortBindings: nat.PortMap{
+			"53/tcp": []nat.PortBinding{
+				{
+					HostIP:   "0.0.0.0",
+					HostPort: "53",
+				},
+			},
+			"53/udp": []nat.PortBinding{
+				{
+					HostIP:   "0.0.0.0",
+					HostPort: "53",
+				},
+			},
+		},
+	}, &network.NetworkingConfig{
+		EndpointsConfig: map[string]*network.EndpointSettings{
+			"coretxtd": &network.EndpointSettings{NetworkID: d.network.ID},
+		},
+	}, "")
+	if err != nil {
+		return fmt.Errorf("Couldn't create the CoreDNS container: %s", err.Error())
+	}
+
+	d.txtdContainer, err = d.cli.ContainerCreate(d.ctx, &container.Config{
+		Image: "okkur/txtdirect:0.4.0",
```

Comment:
We could use `c.txtdirect.org/txtdirect:0.4.0-dirty` here. Also the version of the container should probably be derived from the makefile or else the build container might get out of sync with using it here. What do you think?

Reply:
Yeah, it should definitely read the version from somewhere else like Makefile as you mentioned and not hardcode it. We can set an env variable in `endtoend-test` target for the image name or version and use `os.Getenv()` in here instead.